### PR TITLE
Clean code base and implement MediaRecorder

### PR DESCRIPTION
**Pull request by martinRenou**

### CODE REVIEW

**Comment by martinRenou** [comment]
I took your code from your PR https://github.com/jupyter-widgets/ipywidgets/pull/1685 like you told me.

I plan to do the same for the `WebRTCPeer`, `WebRTCRoom` ... code, maybe on the same PR if it's not too much.

Don't hesitate to tell me if I do something wrong !

File: ipywebrtc/webrtc.py [hunk 1]
```diff
@@ -1,20 +1,30 @@
...
-import ipywebrtc._version
 semver_range_frontend = "~" + ipywebrtc._version.__version_js__
 
-class HasStream:
-    pass # to indicate it has a .stream property on the JS side
-@widgets.register('webrtc.MediaStream')
-class MediaStream(widgets.DOMWidget, HasStream):
-    """Represents a media source."""
+
+@register('webrtc.MediaStream')
```
Comment (maartenbreddels) [inline ipywebrtc/webrtc.py]: We can drop the string, and drop widget 6 support, so just @register should work.

File: ipywebrtc/webrtc.py [hunk 2]
```diff
@@ -23,90 +33,206 @@ class MediaStream(widgets.DOMWidget, HasStream):
...
-    stream = traitlets.Instance(object, allow_none=True).tag(sync=True, **ipywidgets.widget_serialization)
+    stream = Instance(MediaStream, allow_none=True).tag(sync=True, **widget_serialization)
     id = Unicode(read_only=True).tag(sync=True)
     nickname = Unicode('anonymous').tag(sync=True)
-    peers = traitlets.List(traitlets.Instance(WebRTCPeer), [], allow_none=False).tag(sync=True, **ipywidgets.widget_serialization)
-    streams = traitlets.List(traitlets.Instance(MediaStream), [], allow_none=False).tag(sync=True, **ipywidgets.widget_serialization)
+    peers = List(Instance(WebRTCPeer), [], allow_none=False).tag(sync=True, **widget_serialization)
+    streams = List(Instance(MediaStream), [], allow_none=False).tag(sync=True, **widget_serialization)
 
     def close(self):
```
Comment (maartenbreddels) [inline ipywebrtc/webrtc.py]: same here, we can drop close.

**Comment by martinRenou** [comment]
media_recorder

It looks slow on the GIF but it's due to my GIF recorder, and my slow computer... :P

**Comment by maartenbreddels** [comment]
You read my mind implementing this! 👍 
Do we need to show the recording video again btw?

**Comment by martinRenou** [comment]
What do you mean again? With a loop?

**Comment by martinRenou** [comment]
It would be nice to be able to stream from a canvas as well. So we could stream and record ipyvolume or bqplot outputs.

**Comment by maartenbreddels** [comment]
Now I understand, it will only show it again because you do the play button, no I think that is great! We can already record from ipyvolume btw, it's Figure widget inherits from MediaStream. Bqplot is not supported, we might be able to get there if we do some hybrid canvas/svg thing though.

**Comment by martinRenou** [comment]
I totally agree, this PR is big enough :)

**Comment by martinRenou** [comment]
Do you think it would be ready to merge or you would like me to change/add some stuffs ?

**Comment by maartenbreddels** [comment]
I had 2 small fixes I want to commit, but I cannot push onto your branch, can you check if you need to set write permission? I thought by default after you do a PR, I (the repo owner) gets write permission.

**Comment by maartenbreddels** [comment]
The code looks good, everything works, really nice work. I see room for a small improvement:
 * disable play button when recording
 
One bigger issue is that you've removed the .stream property (a Promise) on the model on the JS side (maybe you thought that could be removed because I wanted to get rid of the HasStream part). But it's important that on the JS side we are able to get the MediaStream asynchronously. For instance, this is now broken:
```python
webrtc.chat('bla')
webrtc.chat('bla')
```
You'll see an exception in the notebook.

**Comment by maartenbreddels** [comment]
I don't understand, it still does not work, this seem right:
```
$ git remote -v                                                                  working
martinRenou	https://maartenbreddels@github.com/martinRenou/ipywebrtc.git (fetch)
martinRenou	https://maartenbreddels@github.com/martinRenou/ipywebrtc.git (push)
```

Did you check this: https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

**Comment by martinRenou** [comment]
Concerning the `stream` property, it's now replaced by the `captureStream` method, like you did in your PR. So I guess it fails on this line https://github.com/martinRenou/ipywebrtc/blob/ef99441057f382da288ec06da654a13489d992eb/js/src/webrtc.js#L596 and a few others, the fix would be to replace the `.stream` with `.captureStream()`

**Comment by martinRenou** [comment]
Yes the `Allow edits from maintainers` checkbox is checked by default, so it should work. Why is the url like that? Shouldn't it be https://github.com/martinRenou/ipywebrtc.git ?

**Comment by maartenbreddels** [comment]
That will also break ipyvolume, although I like the change. What about using properties, so that .stream  will invoke captureStream, as a backward compatibility fix?
See https://stackoverflow.com/questions/812961/javascript-getters-and-setters-for-dummies for examples (I had to look it up myself). The defineProperty way may have more browser support I'd guess.

**Comment by maartenbreddels** [comment]
That's just my username, or am I missing something?

**Comment by martinRenou** [comment]
But do you need your username in the url? Did you try without it?

Ok we can do that. If you still can't push on my branch I can make the changes

**Comment by martinRenou** [comment]
Also I added an example directory containing a notebook at the root of the repo, do want me to remove it and change the `docs/source/Local media.ipynb` file? Or should I remove the old one and keep mine?

**Comment by maartenbreddels** [comment]
I had to accept the invite you send me, otherwise I couldn't get it working, plz fetch/merge

**Comment by martinRenou** [comment]
Done :)

**Comment by maartenbreddels** [comment]
I think that notebook file got added accidentally, I'd prefer to have for the docs 1 'tutorial' notebook showing all features, which would also be a good 'test', the only 'issue' is that If we put the chat stuff on it (like on the opening page), if two people open the page they will see eachother.. i'm not sure we want it, but hey, they can always deny webcam access right. So I'd say:
* landing page should be small example, installation instruction
* link to tutorial/demo page showing all parts
* in the future: an api page.

**Comment by maartenbreddels** [comment]
Just to let you know I'm looking into this PR. I actually have trouble setting a media stream as texture for a mesh in ipyvolume, so I cannot say yet if it's working or not. I did notice that the `chat(room='bla')` isn't working btw.

**Comment by maartenbreddels** [comment]
The issue with ipyvolume is solved, I was using a deprecated feature it seems. Only the chat issue stays, do you want to take a look at it?

**Comment by maartenbreddels** [comment]
Nope :(
If you join a room twice from the same notebook, e.g.:
```
r.chat(room='jupytercon')
r.chat(room='jupytercon')
```
It should show yourself 4 times.

**Comment by maartenbreddels** [comment]
Great, it seems to be working. Shall I squash/merge it?
There are no API breaking right, just a patch release? I'd like to get the jupyter lab part up to date, and then do a new release, what do you think?

**Comment by martinRenou** [comment]
There should be no API breaking, I think it should be fine :)
Yep, and once you made a release I will update xwebrtc !

**Comment by maartenbreddels** [comment]
There is one issue left:
 * record = traitlets.Bool(False).tag(sync=True)
Trait on MediaRecorder to control the recording from the backend, i'll see what I can do.

**Comment by maartenbreddels** [comment]
Second: any reason you changed MediaRecords's stream to source? For consistency with other widgets? Otherwise it's a breaking change as well.

**Comment by maartenbreddels** [comment]
See #7b364018501957fb9c778c0ee3d850b88e52f83f and #300d0977b3cc9d23527ec9009d932b5908c49463 for some more changes/fixes I had to make for backwards compatibility

**Comment by martinRenou** [comment]
Maybe I should have made the change about `updateStream()` and `stream` everywhere. The idea of `updateStream()` came from your [PR](https://github.com/jupyter-widgets/ipywidgets/pull/1685/files#diff-dc64f69e62131726181f8f82e3e7f32bR70) on `ipywidgets`, but it would mean changing `ipyvolume` if you rely on the `stream` attribute.
Sorry that it didn't work well for you

**Comment by maartenbreddels** [comment]
No problem, it was indeed more elegant, should have done that from the start, we now have to live with it:)

**Pull request merged**